Topic wireframes issue#2

jkluse opened the pull request:
Ready for some feedback on the wireframes

Only difference between Add Topic/Edit Topic pages is that Add Topic doesn't have a status input field (since a new topic would be Active by default).

Let me know if I missed and fields or if you'd like to modify the design.

FlatBallFlyer commented on docs/wireframes/add-topic-wireframe.png:
How are you going to manage auto-save functions on this page?

jkluse replied:
Could you describe the auto-save feature you would like? Is this something that would save on the front-end or something that will interact with the backend POST API?

FlatBallFlyer replied:
See comments below

FlatBallFlyer commented on docs/wireframes/edit-topic-wireframe.png:
Can we use a [+] plus button on the collapse header instead of separate buttons?
Can you please wire frame the skills and resources section?
I think the "Add Topic" button on this page needs housekeeping.

jkluse replied:
1. So you would like a + button in lieu of a named button for Add Skill and Add Resource, is that right? 
2. You are asking for the wireframes for how the Add Skill and Add Resource page or modal would look? 
3. Not sure what you mean by housekeeping. I will remove the blue color. Was there something else?

FlatBallFlyer replied:
1. Yes - but I'm open
2. Yes please
3. You have an Add Topic button at the bottom of both the New and Edit pages

FlatBallFlyer commented on docs/wireframes/topics-wireframe.png:
I think it would be valuable to have the category/topic relationship be exposed on this page. Maybe accordions by Category, with the below list of topics in each. The list of valid categories is in the enums. I do not anticipate any capacity performance issues on page, but if this collection gets huge we can lazy load and paginate topics in a category.

jkluse replied:
That makes sense, I can wireframe that concept with accordion categories.

jkluse replied:
Something like this? Would we need to update the /api/topic GET endpoint to return category? At present, it appears to just return id and name.

<img width="200" alt="topics-wireframe" src="https://github.com/agile-learning-institute/mentorHub-topic-ui/assets/25851301/555aadb4-93e0-4a58-be0a-53a58201f120">

FlatBallFlyer replied:
We can either
- MVP - update the get endpoint to return grouped by results 
- Use the enums to populate the top level categories, then implement a query parameter on the GET API endpoint that specifies the category to get, then lazy load the list in the drop down when it is opened

FlatBallFlyer reviewed (changes requested):
Looking good! Feedback left on specific wireframes, but here are a few nit's 
- For now we are using very generic looking content, blue isn't bad but the person-ui used grey pretty consistently, just to keep is super generic. 
- Have you thought about the user stories we will use for testing the UI? It would be good to review that before you invest too much in code.  Seems like a good use of markdown. 
- I'm looking forward to seeing how you will handle skills and resources.

I was looking at Path and Sequence and thinking we might want to create a list view for Paths with drag-and-drop sequencing instead of exposing sequence directly. We have 2 ways of grouping content - Category and Path - Paths typically cross multiple categories. Maybe 2 list pages, or maybe a list page that has buttons or tabs, I'm open to ideas here.

FlatBallFlyer commented on docs/wireframes/admin-wireframe.png:
I don't think you need Partners or Mentors

jkluse replied:
I will update wireframe

FlatBallFlyer commented:
Re: ^^

1. Yes, please create user stories, I doubt there will be much there it's a very simple UI. There should be enough to cover pretty path testing of all functionality. 

2. Let's prototype it, in my mind it was just a form exposed in the accordion. In general I'm not a fan of modal dialogs. 

3. Sequence from the data perspective is just any value used for sorting. If I had a list of topics within a Path and I can drag items in the list to re-arrange them then the UI can adjust sequence under the covers. See [here](https://jqueryui.com/sortable/) for an example of a draggable/sortable list. (NO I'M NOT PROPOSING JQUERY)

4. Auto-Save - on-change (on-blur) to any property, make an API patch call with only the property that changed.  The full updated document will be returned, so there is a pseudo-real-time-concurrency component to the pattern as well. See the person-ui for an example. - More specifically about the New Topic page, you only need a name on the initial POST api call, and you MUST have a unique name on that call, so only ask for that one value, after a successful POST you just switch over to the edit view.

jkluse commented:
@FlatBallFlyer  I added some basic user stories, and updated all of the wireframes. Ready for another round of feedback. 


I haven't updated sequences yet because I wanted to clarify a couple more things. So the drag-and-drop sequence topics would displayed on the `/topics/` page under the respective category. Is that correct? So based on this feature, would the edit topic page display a field for sequence or just default to put it at the end? Would using a number for sequence field be easier than a string for sorting?

FlatBallFlyer commented:
It's looking good. Here is some more feedback:

# Status
Since we only have 2 statuses (Active and Archived) let's just have a simple button, at the bottom of the form. If the status is "Active" the button is "Archive" and it changes the status to Archived. If the status is Archived the button says "Restore" and it changes the status to Active. When we implement RBAC this control will only be available for users with the "Staff" role. 

# Topic List view and Path Sequence
It seems to me like we need 2 list views (grouped by Category and grouped by Path). On the group-by-path view the items within a path can be dragged to a different place in the same path (Not between paths). When dragged the UI is responsible for adjusting sequence on as many topics as required and persisting those changes. The sequence would not be exposed in the edit topic view. If you want to change sequence to be numeric I'm fine with that, to do that you should open an issue on the mongodb repo, and the topic-api repo to make that change. 

# User Stories
I think of user stories slightly differently than you, I'm open to different ideas, but I like to use the Design Thinking approach to user stories. Every user story should start with a Needs statement, and then lists a series of steps... so something like this:

## Mentors need a way to find a specific Topic by Category
- Opens /
- Redirects to /topics with by category layout
- Clicks on Topic and opens /topic view

## Mentors need a way to add skills to a topic
- Opens /topic/OID
- Clicks on + 
- Fills in new Skill

## Mentors need a way to remove a skill from a topic
- Opens /topic/OID
- Expands Skills section
- Clicks trashcan next to skill
- Confirms delete
- Form updates

It is also nice to have a slug to identify the need/story, so find a way to include that in the Markdown - maybe on the ## Line as in

## add-skill: Mentors need a way to add a skill to a Topic

FlatBallFlyer commented:
Also thinking about the add/edit a resource form - we could open that in a new page.... I don't know, it might be nice to be able to edit multiple resources at the same time (multiple tabs), and we might have people who want to view/edit a resource without navigating to the topic. What are your thoughts about this approach?

jkluse commented:
@FlatBallFlyer  Regarding your comments above, I like your idea and think it is cleaner to have the add and edit a resource on a separate page (wouldn't need to insert a form into the edit topic page). There could also be a link to 'Add a Resource' on the `/topics` page. The same could be done for skills as well. However, if we implemented this, we would need a new list, add, and edit route for resources (and skills), right?

FlatBallFlyer commented:
I'm thinking that:
- We need to have a /resource/[id] path that edits a resource. This will likely be used from other UI's that reference resources directly. 
- The add resource button in the Topic page does the post before hading off to /resource/[id] for editing. The Topic needs the resource ID. 
- No need for a list resource or skill UI - the Topic lists these
- I think probably skills is the same, but it's such a simple object that maybe we do it in-place in the Topic.... I'm less sure about this part of the design.

jkluse commented:
@FlatBallFlyer 

Pushed commit to this PR branch:
- updated format for user-stories
- created two different wireframes for topics with categories and paths
- path view now has a path/sub-path dropdown to handle discussed changes to path hierarchy

Let me know your feedback, thanks!

FlatBallFlyer reviewed (approved):
I think this is good enough to move ahead with building out the UI. I think we could have done a much better job of design-thinking this, have a collaborative workshop, involved more than just me as the sole sponsor user, experiment with paper prototypes before building clean wireframes, etc. Let's remember that for the next UI. 

I do think there will be some new user stories from segments... mentor needs a way to create a new Path, a new Segment, re-order Segments within a Path, re-order topics within a Segment.... go ahead and draft those up as we move through, but I think we are ready to proceed to setting up the project. You can merge this to main to close the PR,  you can continue to work in this branch for the next step if you want to, or delete the branch after the merge.   

For the next step, review the gitHub Actions, Dockerfile, nginx proxy config, and the npm proxy config in the [person-ui](https://github.com/agile-learning-institute/mentorHub-person-ui) repo. When you have a minimally functional web site (hello-world for every route) open a new PR. I will need to manually build and push the container the first time, after that the CI should handle it.
